src/app-layer-frames.c

```diff
@@ -0,0 +1,416 @@
+{
+    memcpy(dst, src, sizeof(*dst));
+}
+
+static void AppLayerFrameDumpForFrames(const char *prefix, const Frames *frames)
+{
+    // uint32_t last_re = 0;
+    for (uint16_t i = 0; i < frames->cnt; i++) {
+        if (i < FRAMES_STATIC_CNT) {
+            const Frame *frame = &frames->sframes[i];
+            FrameDebug(prefix, frames, frame);
+            // BUG_ON(last_re != 0 && last_re > frame->rel_offset);
+            // last_re = frame->rel_offset + frame->len;
+        } else {
+            const uint16_t o = i - FRAMES_STATIC_CNT;
+            const Frame *frame = &frames->dframes[o];
+            FrameDebug(prefix, frames, frame);
+            // BUG_ON(last_re != 0 && last_re > frame->rel_offset);
+            // last_re = frame->rel_offset + frame->len;
+        }
+    }
+}
+
+/** Stream buffer slides forward, we need to update and age out
+ *  frame offsets/frames. Aging out means we move existing frames
+ *  into the slots we'd free up.
+ *
+ *  Start:
+ *
+ *  [ stream ]
+ *    [ frame   ...........]
+ *      rel_offset: 2
+ *      len: 19
+ *
+ *  Slide:
+ *         [ stream ]
+ *    [ frame ....          .]
+ *      rel_offset: -10
+ *       len: 19
+ *
+ *  Slide:
+ *                [ stream ]
+ *    [ frame ...........    ]
+ *      rel_offset: -16
+ *      len: 19
+ */
+int FrameSlide(Frames *frames, uint32_t slide)
+{
+    BUG_ON(frames == NULL);
+    SCLogDebug("frames %p: sliding %u bytes", frames, slide);
+
+    if (slide >= frames->progress_rel)
+        frames->progress_rel = 0;
+    else
+        frames->progress_rel -= slide;
+
+    uint16_t x = 0;
+    for (uint16_t i = 0; i < frames->cnt; i++) {
+        if (i < FRAMES_STATIC_CNT) {
+            Frame *frame = &frames->sframes[i];
+            FrameDebug("slide(s)", frames, frame);
+            if (frame->rel_offset + frame->len <= (int32_t)slide) {
+                // remove by not incrementing 'x'
+                SCLogDebug("removing %p", frame);
+                FrameClean(frame);
+            } else {
+                Frame *nframe = &frames->sframes[x];
+                StreamDPUCopy(nframe, frame);
+                nframe->rel_offset -= slide; /* turns negative if start if before window */
+                if (frame != nframe) {
+                    FrameClean(frame);
+                }
+                x++;
+            }
+        } else {
+            const uint16_t o = i - FRAMES_STATIC_CNT;
+            Frame *frame = &frames->dframes[o];
+            FrameDebug("slide(d)", frames, frame);
+            if (frame->rel_offset + frame->len <= (int32_t)slide) {
+                // remove by not incrementing 'x'
+                SCLogDebug("removing %p", frame);
+                FrameClean(frame);
+            } else {
+                Frame *nframe;
+                if (x >= FRAMES_STATIC_CNT) {
+                    nframe = &frames->dframes[x - FRAMES_STATIC_CNT];
+                } else {
+                    nframe = &frames->sframes[x];
```

Comment:
Q: Why do we set a static frame here if we are in the dynamic frame zone?

Reply:
we're moving this after removing frames from our list, so we can move frames from dynamic to static